ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -89,9 +94,12 @@ public void run() {
         handle = txnHandler.getMutexAPI().acquireLock(TxnStore.MUTEX_KEY.Cleaner.name());
         startedAt = System.currentTimeMillis();
         long minOpenTxnId = txnHandler.findMinOpenTxnIdForCleaner();
+        List<CompletableFuture> cleanerList = new ArrayList<>();
         for(CompactionInfo compactionInfo : txnHandler.findReadyToClean()) {
-          clean(compactionInfo, minOpenTxnId);
+          cleanerList.add(CompletableFuture.runAsync(ThrowingRunnable.unchecked(() ->
+            clean(compactionInfo, minOpenTxnId)), cleanerExecutor));
         }
+        CompletableFuture.allOf(cleanerList.toArray(new CompletableFuture[0])).join();
```

Comment:
What happens if one of the futures throws the exception?
Are the others continue to execute? Will we wait until all of the tasks are finished one way or another?
We do not want multiple Cleaning tasks running concurrently on the same partition.

Seeing this the same problem might arise with the Initiator too. What do you think @deniskuzZ ?

Reply:
Talked with @deniskuzZ and he convinced me that join() will wait for all of the tasks to finish :)